includes/classes/class-rtcamp-google-embeds.php

```diff
@@ -261,15 +270,17 @@ private function get_thumbnail_url( $file_id, $user_id = false ) {
 		// Check if a preview exists for supplied file id.
 		$thumbnail_url = sprintf( 'https://drive.google.com/thumbnail?id=%s&sz=w400-h400', $file_id );
 		$response      = wp_remote_get( $thumbnail_url );
+		$contents      = wp_remote_retrieve_body( $response );
 		if ( ! is_wp_error( $response ) ) {
-			
 			// Check if retrieved content is image and not google sign up page.
 			$content_type = wp_remote_retrieve_header( $response, 'content-type' );
 			if ( false !== strpos( $content_type, 'image/' ) ) {
-
 				// Check if retrieved http code is 200.
 				$status_code = wp_remote_retrieve_response_code( $response );
 				if ( 200 === $status_code ) {
+					// Save the thumbnail.
+					$this->save_thumbnail( $file_id, $contents );
```

Comment:
You need to cache image which is being fetched on line 322 of this file too. And test once if it's working fine even after token is expired

Reply:
Did you try implementing this? Current `save_thumbnail` call will only work for public files, not for private files. In order for this to work with private files, you'll need to do what I've suggested above.

Reply:
Also `get_thumbnail_url` function should return cached URL if network request fails. If all the network requests fails, then at the end you can check whether the file of this ID is already cached, if it is then send that URL.

Reply:
@SID177 The idea is if a cached image is saved, it will show the preview from the cached image. Otherwise, it will pull the preview image and save it and then show the preview from the saved image. Not sure why network request failing or token expiration is issue.